### ISSUE TITLE
[libretro] Option to disable VMU 1 (or emulate MU instead) for Sonic Shuffle

### ISSUE THREAD

**Issue by GHNeko** [flyinghead/flycast#723]
**Please Note: This form is the minimum required information for submitting bugs.**  
**Removing this form may lead to your issue being closed until it is completed.**

Platform / OS / Hardware: RetroArch + Flycast Core / Windows 8.1 & Arch Linux

Github hash: 

Hardware: N/A

**_Description of the Issue_**

I was sent here by bslenul after being corrected and told that this is a Flycast issue.

"Yes but both the standalone and the core are maintained from the repo I linked, so even if it's an issue that only affects RetroArch it should be posted there." as per bslenul from the libretro/RetroArch github.

Here is the issue as copied from RetroArch

Sonic Shuffle is a party game for the Sega Dreamcast. It's a first party game that makes heavy use of the VMU, but Sega realized that not everyone would have their LCD VMUs and coded behaviors that trigger based on what device is in slot 1 and if there is a device in slot 1.

Generally in VS mode, when you load into the game with a VMU in Slot 1, when the player is selecting cards, their hand is shown on their VMU screen instead of in game. However if a normal memory card is in slot 1 (no screen) or there is no device in slot 1 (VMU in slot 2) the game will instead display the player's hand on the screen.

This is important because while this is a non issue for single player play (as you can emulate the vmu display), for multiplayer purposes, emulating VMU display essentially is showing all players hands to everyone and that's not how you play this primarily multiplayer game lol.

If you do not emulate the VMU display, but the game still sees a VMU instead of an MU in slot 1, it will still hide the cards on screen and then as a result you cannot see what card you have, breaking a lot of the functionality of the game for multiplayer as the game's design is focused around knowing what you have and it is designed to be a game that's played with friends.

So tldr - Because RetroArch only emulates a VMU and not also an MU, and always has a VMU in Slot 1 of Controller A or Controllers A-D; you cannot trigger the contingency for when the game/console sees that there is either an incompatible device or no device in slot 1.

To further supplement this idea, if you play Story Mode (single player only experience), the players cards are always face up regardless of what's in Slot 1 as by design its a single player experience and there is no issue in having your cards visible.

**_Debugging Steps Tested_**

  * Fill in any steps already tried here
  * Begin each new line with an asterisk
  * If no steps are required, please skip

**_Logs Gathered_**

```

Please paste the contents of the log / logs here
You may leave this field blank if you have none.

```

**_Screenshots_**

(Replace this line with any issue screenshots)

**Comment by flyinghead** [comment]
I don't see how emulating MU (VMU without LCD) is going to solve the issue. Flycast only has one output screen, where both players' hands will be displayed, either by the game or as a VMU screen. In both cases each player sees the other player's hand.

Obviously this limits the interest of this game mode but I don't see any reasonable solution.

**Comment by GHNeko** [comment]
Emulating MU will cause the game to trigger its intended behavior which is to display the cards on the screen.

Sent via the Samsung Galaxy Z Fold3 5G, an AT&T 5G smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by bslenul** [comment]
This is how it works:

* If a VMU is in slot 1 the cards are hidden on-screen:

![image](https://user-images.githubusercontent.com/33353403/187450587-72f99917-3c6d-41f1-8e0e-1f65430edbf4.png)

* If no VMU in slot 1 the cards are shown:

![image](https://user-images.githubusercontent.com/33353403/187450741-9f9c053a-fca7-42dc-9e65-ab3d0a68cc98.png)

If I understand the issue correctly, this is the problem:

In standalone as a workaround you can simply disable slot 1 VMUs and use slot 2 instead, then ask other players to look away when it's your turn to pick a card.

However with the Libretro core the cards will ALWAYS be hidden on-screen for player 1 (or for all players depending on your VMU core options), and if you enable the OSD for the VMUs they'll be shown at all time so you don't want that either.

Being able to emulate MUs or being able to use slot 2 VMUs instead of slot 1 would fix the issue with the core.

**Comment by GHNeko** [comment]
Yeah I apologize for not clarifying that this is Libretro specific. Technically redream has the same issue but that's that.

Flycast has a ggpo branch but that is only for 2 players and that likely won't change.

Libretro has built in Netplay and streamlines the process of playing games via Netplay and since the game is way worse to play because of this I wanted to see if it could be addressed.

Sent via the Samsung Galaxy Z Fold3 5G, an AT&T 5G smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by bslenul** [comment]
I'm messing with this since yesterday, I made it so that you can configure each slot for each device from the core options (with options being hidden if needed depending on the device type):

https://user-images.githubusercontent.com/33353403/187733330-4cd63fa4-2cc4-4bef-a674-b5954316a930.mp4

Works fine with Sonic Shuffle, after "unplugging" the A1 VMU the cards now appear on-screen:

https://user-images.githubusercontent.com/33353403/187734625-85f3a57b-9144-41b1-80c0-3c78444ca250.mp4

I don't feel comfortable sending a PR tho, I'm not a dev so code-wise it's probably ugly AF :/ But from my tests it works pretty well. Changes are here: https://github.com/bslenul/flycast/commit/14ce2265d720c2ddd00e860acb3c35778c076d84

Also I put the options in "Input" category, idk if it'd be better to move them in "System" or maybe create a "Ports Expansions" category...

But maybe Flyinghead as something completely different in mind to handle this anyway :p

**Comment by GHNeko** [comment]
Yeah this is pretty much the implimentation I was hoping for from a functionality pov. Thank you for the work and hopefully this is acceptable for Flyinghead.

**Comment by bslenul** [comment]
Updated (see the edit from my previous reply), if you want to give it a try: [flycast_windows.zip](https://github.com/flyinghead/flycast/files/9547568/flycast_windows.zip) and/or [flycast_linux.tar.gz](https://github.com/flyinghead/flycast/files/9547724/flycast_linux.tar.gz)

It shouldn't mess with your saves at all, but just to be extra careful you might want to backup your VMUs first.

**Comment by GHNeko** [comment]
Ill try it out asap. Currently migrating from windows to Linux as a daily driver so Thanks for providing both binaries lol

Sent via the Samsung Galaxy Z Fold3 5G, an AT&T 5G smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by MichaelTheriot** [comment]
> Updated (see the edit from my previous reply), if you want to give it a try: [flycast_windows.zip](https://github.com/flyinghead/flycast/files/9547568/flycast_windows.zip) and/or [flycast_linux.tar.gz](https://github.com/flyinghead/flycast/files/9547724/flycast_linux.tar.gz)
> 
> It shouldn't mess with your saves at all, but just to be extra careful you might want to backup your VMUs first.

This is awesome, thanks! I had the same issue with Sonic Shuffle and this resolves it.

This also completely solved another issue I had with using extra VMUs since it is now trivial to eject, move the files on my machine around, then insert again. Before I was creating save states and restarting each time I had to perform this.

For what it's worth, I instinctively went to the Input menu to search for a feature like this before finding this patch, but I think the Controller Expansion Slots menu is also intuitive now that I see it.

**Comment by bslenul** [comment]
Thank you for testing ❤️ I just wish I was more confident about this to make an actual PR 😓 

And yeah, I wasn't sure about the category thing, but I thought it felt weird to go in "Input" to un/plug VMUs (although it kinda makes sense on Dreamcast).

**Comment by blueminder** [comment]
> Flycast has a ggpo branch but that is only for 2 players and that likely won't change. 

Netplay functionality hasn't been in its own branch for a long time. It's a mainline feature at this point.

I also got 4 player working over GGPO in Flycast Dojo and hope to add the changes upstream after releasing it. I'm just fixing some replay details at this point before that happens.

https://www.youtube.com/watch?v=KdFxpg2x9Ns

**Comment by GHNeko** [comment]
Yeah you had told me on Twitter that you were working on supporting commercial games! I'll be waiting on that as well.

Sent via the Samsung Galaxy Z Fold3 5G, an AT&T 5G smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by blueminder** [comment]
As far as this feature goes, there's no real distinction between commercial or homebrew games.

**Comment by GHNeko** [comment]
Roger that! Thanks for the clarification then. I'm still looking forward to either the patreon release or public release.

**Comment by gamefb** [comment]
Is there any news as to whether or not something like this will get pushed?
Also, how would the patch get added if I already have flycast installed through retroarch/EmuDeck?

**Comment by mattpoulin789** [comment]
I wish that the flycast emulator dlc was updated on the steam version of retroarch to include the Controller Expansion Slots menu so whenever you install the flycast emulator dlc for retroarch on steam it will have the Controller Expansion Slots menu to be able to enable and disable the VMUs in the steam version of the flycast emulator
